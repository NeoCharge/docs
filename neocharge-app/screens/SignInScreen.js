import React from 'react';
import { View, Text, TextInput, Button, StyleSheet, AsyncStorage } from 'react-native';
import { Auth } from 'aws-amplify';
import * as SecureStore from 'expo-secure-store';

class SignInScreen extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            EmailInputValue: '',
            PasswordInputValue: '',
            ErrorMessage: '',
            jsonDeviceLogs: []
        }
    }
  
    render() {
      return (
        <View style={styles.screen}>
            <View style={styles.contents}>
                <Text style={styles.title}>Sign-In</Text>
                <Text style={styles.ErrorText}>{this.state.ErrorMessage}</Text>
                <TextInput
                    style={styles.inputContainer}
                    placeholder='Email'
                    onChangeText={EmailInputValue => this.setState({ EmailInputValue })}
                    autoCapitalize='none'
                />
                <TextInput
                    style={styles.inputContainer}
                    placeholder='Password'
                    onChangeText={PasswordInputValue => this.setState({ PasswordInputValue })}
                    secureTextEntry={true}
                    autoCapitalize='none'
                />

                <Button title="Sign in" 
                onPress={() => this.SignIn()} />
                
                <Text 
                    style={styles.ClickableText} 
                    onPress={ ()=>  this.props.navigation.navigate('SignUp')}>
                    Don't have an account? Click here to Sign-Up.
                </Text>
            </View>
        </View>
      );
    }


    SignIn = async() => {
        const email = this.state.EmailInputValue;
        const password = this.state.PasswordInputValue;
        var noErrors = true;
        try {
            if(!(noErrors = this.checkValidInput(email, password))) {
                return //return if the input was not valid
            }
            const user = await Auth.signIn(email, password)
                .catch(error => {
                    console.log(error.code);
                    noErrors = false;
                    this.handleErrors(error.code, email);
                });
            console.log(user);
            if (noErrors) {
                this.setSecureStore("secure_email", email);
                this.setSecureStore("secure_password", password);
                this.props.navigation.navigate('App');
            }
        } catch (err) {
            console.log("catching error: " + err);
        }
    }


    checkValidInput(email, password) {
        if (email.length === 0) {
            console.log("No email was given.");
            this.setState({ErrorMessage: "Please enter an email address."});
            return false;
        } 
        if (password.length === 0) {
            console.log("No password was given.");
            this.setState({ErrorMessage: "Please enter your password."});
            return false;
        }
        return true;
    }


    handleErrors(errorcode, email) {
        if (typeof(errorcode) === "undefined") {
            console.log("Email and password cannot be empty.");
            this.setState({ErrorMessage: "Email and password cannot be empty."});
        } 
        else if (errorcode === 'UserNotConfirmedException') {
            // The error happens if the user didn't finish the confirmation step when signing up
            console.log("This account has not yet been verified.");
            this.setState({ErrorMessage: "This account has not yet been verified."});
            this.props.navigation.navigate('Verify', {userEmail : email});
        } 
        else if (errorcode === 'PasswordResetRequiredException') {
            // The error happens when the password is reset in the Cognito console
            // In this case you need to call forgotPassword to reset the password
            console.log("This account's password has been reset.");
            this.setState({ErrorMessage: "Your account's password has been reset. Please click \"Forgot Password\" to create a new password."});
        } 
        else if (errorcode === 'NotAuthorizedException') {
            // The error happens when the incorrect password is provided
            console.log("Incorrect Password");
            this.setState({ErrorMessage: "Incorrect Password"});
        } 
        else if (errorcode === 'UserNotFoundException') {
            // The error happens when the supplied username/email does not exist in the Cognito user pool
            console.log("Specified user could not be found.");
            this.setState({ErrorMessage: "The specified user could not be found. Please try another email."});
        } 
        else if (errorcode === 'NetworkError') {
            console.log("Network error.");
            this.setState({ErrorMessage: "Network error."});
        } else {
            console.log("Something else went wrong");
            this.setState({ErrorMessage: "Something else went wrong."});
        }
    }

    setSecureStore = async (key, value) => {
        console.log("storing key: " + key + ", value: " + value);
        await SecureStore.setItemAsync(key, value);
    }
}
export default SignInScreen;

const styles = StyleSheet.create({
    screen: {
        padding: 30,
        backgroundColor: "#242424", //dark gray
        flex: 1,
    },
    contents: {
        top: '20%',
        bottom: '20%',
        alignItems: 'center',
        flexDirection: 'column',
        justifyContent: 'center',
    },
    title: {
        color: '#fff', //white
        fontSize: 20,
        marginBottom: '5%',
    },
    ErrorText: {
        color: '#ff0000', //red
        flexDirection: 'column',
        marginBottom: '5%',
    },
    inputContainer: {
        height: 40,
        width: '80%',
        color: 'white',
        borderColor: 'gray',
        paddingLeft: 10,
        borderWidth: 1,
        marginBottom: '5%',
    },
    ClickableText: {
        color: '#E88227', //oragne
        textDecorationLine: 'underline',
        marginTop: '10%',
    },
});

