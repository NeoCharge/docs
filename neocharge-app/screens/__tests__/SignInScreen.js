/**
 * @jest-environment jsdom
 */
import 'react-native';
import React from 'react';
import SignInScreen from '../SignInScreen';
import { Auth } from 'aws-amplify';
import { mount, shallow, } from 'enzyme';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {
	CookieStorage,
	CognitoUserPool,
	CognitoUser,
	CognitoUserSession,
	CognitoIdToken,
	CognitoAccessToken,
} from 'amazon-cognito-identity-js';

const idToken = new CognitoIdToken({ IdToken: 'idToken' });
const accessToken = new CognitoAccessToken({ AccessToken: 'accessToken' });

const session = new CognitoUserSession({
	IdToken: idToken,
	AccessToken: accessToken,
});
/*
const user = new CognitoUser({
  Username: 'username',
  Pool: userPool,
});
*/

const signin_screen = require('../SignInScreen');

describe('snapshot testing', () => {
  test('renders correctly', () => {
    const tree = renderer.create(<SignInScreen />).toJSON();
      expect(tree).toMatchSnapshot();
  });
});

describe('signIn', () => {
  test('successful case', async () => {

    const component = new SignInScreen();

    // Force valid input
    const spyon1 = jest
      .spyOn(component, 'checkValidInput')
      .mockImplementation((authenticationDetails, callback) => {
        return true;
      });

    // Force successful signin
    const spyon2 = jest
      .spyOn(Auth, 'signIn')
      .mockImplementation((authenticationDetails, callback) => {
        callback.onSuccess(session);
      });

    // Spy on error handling to verify it doesn't run
    const spyon3 = jest
      .spyOn(component, 'handleErrors');

    // Call function for testing
    component.setState({EmailInputValue: 'email'});
    component.setState({PasswordInputValue: 'Password1'});
    await component.SignIn();

    // Check expected behavior
    expect(spyon1).toBeCalled();
    expect(spyon2).toBeCalled();
    expect(spyon3).not.toHaveBeenCalled();
    expect.assertions(3);

    spyon1.mockClear();
    spyon2.mockClear();
    spyon3.mockClear();
  });

  test('invalid input case', async () => {

    const component = new SignInScreen();

    // Force invalid input
    const spyon1 = jest
      .spyOn(component, 'checkValidInput')
      .mockReturnValue(false);

    // Spy on signin to verify it doesn't run
    const spyon2 = jest
      .spyOn(Auth, 'signIn')

    // Spy on error handling to verify it doesn't run
    const spyon3 = jest
      .spyOn(component, 'handleErrors');

    // Call function for testing
    component.setState({EmailInputValue: ''});
    component.setState({PasswordInputValue: ''});
    await component.SignIn();

    // Check expected behavior
    expect(spyon1).toBeCalled();
    expect(spyon2).not.toHaveBeenCalled();
    expect(spyon3).not.toHaveBeenCalled();
    expect.assertions(3);

    spyon1.mockClear();
    spyon2.mockClear();
    spyon3.mockClear();
  });

  test('failed authentication case', async () => {

    const component = new SignInScreen();

    // Force valid input
    const spyon1 = jest
      .spyOn(component, 'checkValidInput')
      .mockImplementation((authenticationDetails, callback) => {
        return true;
      });

    // Throw error on signin
    const spyon2 = jest
      .spyOn(Auth, 'signIn')
      .mockImplementation((authenticationDetails, callback) => {
        throw new Error();
      });

    // Spy on error handling to verify it runs
    const spyon3 = jest
      .spyOn(component, 'handleErrors');

    // Call function for testing
    component.setState({EmailInputValue: 'email'});
    component.setState({PasswordInputValue: 'Password1'});
    await component.SignIn();

    // Check expected behavior
    expect(spyon1).toBeCalled();
    expect(spyon2).toBeCalled();
    expect(spyon3).toBeCalled();
    expect.assertions(3);

    spyon1.mockClear();
    spyon2.mockClear();
    spyon3.mockClear();
  });


});