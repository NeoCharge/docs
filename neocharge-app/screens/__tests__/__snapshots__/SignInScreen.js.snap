// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#242424",
      "flex": 1,
      "padding": 30,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": "20%",
        "flexDirection": "column",
        "justifyContent": "center",
        "top": "20%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 20,
          "marginBottom": "5%",
        }
      }
    >
      Sign-In
    </Text>
    <Text
      style={
        Object {
          "color": "#ff0000",
          "flexDirection": "column",
          "marginBottom": "5%",
        }
      }
    >
      
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Email"
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "color": "white",
          "height": 40,
          "marginBottom": "5%",
          "paddingLeft": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "color": "white",
          "height": 40,
          "marginBottom": "5%",
          "paddingLeft": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Sign in
        </Text>
      </View>
    </View>
    <Text
      onPress={[Function]}
      style={
        Object {
          "color": "#E88227",
          "marginTop": "10%",
          "textDecorationLine": "underline",
        }
      }
    >
      Don't have an account? Click here to Sign-Up.
    </Text>
  </View>
</View>
`;

exports[`snapshot testing renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#242424",
      "flex": 1,
      "padding": 30,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": "20%",
        "flexDirection": "column",
        "justifyContent": "center",
        "top": "20%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 20,
          "marginBottom": "5%",
        }
      }
    >
      Sign-In
    </Text>
    <Text
      style={
        Object {
          "color": "#ff0000",
          "flexDirection": "column",
          "marginBottom": "5%",
        }
      }
    >
      
    </Text>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Email"
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "color": "white",
          "height": 40,
          "marginBottom": "5%",
          "paddingLeft": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "borderColor": "gray",
          "borderWidth": 1,
          "color": "white",
          "height": 40,
          "marginBottom": "5%",
          "paddingLeft": 10,
          "width": "80%",
        }
      }
      underlineColorAndroid="transparent"
    />
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Sign in
        </Text>
      </View>
    </View>
    <Text
      onPress={[Function]}
      style={
        Object {
          "color": "#E88227",
          "marginTop": "10%",
          "textDecorationLine": "underline",
        }
      }
    >
      Don't have an account? Click here to Sign-Up.
    </Text>
  </View>
</View>
`;
